src/mqtt_client.rs

```diff
@@ -159,64 +166,58 @@ impl<'a> Command<'a> {
 where
     Settings: TreeKey<Y> + Clone,
     Stack: TcpClientStack,
     Clock: embedded_time::Clock,
+    Broker: minimq::Broker,
 {
-    mqtt: minimq::Minimq<Stack, Clock, MESSAGE_SIZE, 1>,
+    mqtt: minimq::Minimq<'buf, Stack, Clock, Broker>,
     settings: Settings,
     state: sm::StateMachine<sm::Context<Clock, Settings, Y>>,
     prefix: String<MAX_TOPIC_LENGTH>,
-    listing_state: Option<Iter<Settings, Y>>,
-    properties_cache: Option<Vec<u8, MESSAGE_SIZE>>,
-    pending_response: Option<Response<32>>,
+    listing_state: Option<(String<MAX_TOPIC_LENGTH>, Iter<Settings, Y>)>,
 }
 
-impl<Settings, Stack, Clock, const MESSAGE_SIZE: usize, const Y: usize>
-    MqttClient<Settings, Stack, Clock, MESSAGE_SIZE, Y>
+impl<'buf, Settings, Stack, Clock, Broker, const Y: usize>
+    MqttClient<'buf, Settings, Stack, Clock, Broker, Y>
 where
     for<'de> Settings: JsonCoreSlash<'de, Y> + Clone,
     Stack: TcpClientStack,
     Clock: embedded_time::Clock + Clone,
+    Broker: minimq::Broker,
 {
     /// Construct a new MQTT settings interface.
     ///
     /// # Args
     /// * `stack` - The network stack to use for communication.
-    /// * `client_id` - The ID of the MQTT client. May be an empty string for auto-assigning.
     /// * `prefix` - The MQTT device prefix to use for this device.
-    /// * `broker` - The IP address of the MQTT broker to use.
     /// * `clock` - The clock for managing the MQTT connection.
     /// * `settings` - The initial settings values.
+    /// * `config` - The configuration of the MQTT client.
     pub fn new(
         stack: Stack,
-        client_id: &str,
         prefix: &str,
-        broker: IpAddr,
         clock: Clock,
         settings: Settings,
+        config: minimq::Config<'buf, Broker>,
     ) -> Result<Self, minimq::Error<Stack::Error>> {
-        let mut mqtt = minimq::Minimq::new(broker, client_id, stack, clock.clone())?;
-
-        // Note(unwrap): The client was just created, so it's valid to set a keepalive interval
-        // now, since we're not yet connected to the broker.
-        mqtt.client()
-            .set_keepalive_interval(KEEPALIVE_INTERVAL_SECONDS)?;
-
-        let prefix = String::from(prefix);
-
         // Configure a will so that we can indicate whether or not we are connected.
+        let prefix = String::from(prefix);
         let mut connection_topic = prefix.clone();
         connection_topic.push_str("/alive").unwrap();
-        mqtt.client().set_will(
-            &connection_topic,
-            b"0",
-            QoS::AtMostOnce,
-            Retain::Retained,
-            &[],
-        )?;
+        let mut will = minimq::Will::new(&connection_topic, b"0", &[]).unwrap();
+        will.retained(Retain::Retained);
+        will.qos(QoS::AtMostOnce);
+
+        let mut config = config
+            .keepalive_interval(KEEPALIVE_INTERVAL_SECONDS)
+            .autodowngrade_qos();
+
+        config.will(will).ok();
```

Comment:
Not sure what the best choice of action here is if the user doesn't provide us with `will_buffer` memory in the MQTT client config.

Reply:
Current plan is to refactor the will buffer management.